src/main/java/com/iexec/worker/chain/IexecHubService.java

```diff
@@ -175,6 +177,65 @@ private IexecHubContract.TaskRevealEventResponse sendRevealTransaction(String ch
     }
 
+    public Optional<ChainReceipt> contributeAndFinalize(Contribution contribution, String resultLink, String callbackData) {
+        try {
+            return CompletableFuture.supplyAsync(() -> {
+                log.info("contributeAndFinalize request [chainTaskId:{}, waitingTxCount:{}]",
+                        contribution.getChainTaskId(), getWaitingTransactionCount());
+                IexecHubContract.TaskFinalizeEventResponse finalizeEvent = sendContributeAndFinalizeTransaction(contribution, resultLink, callbackData);
+                return Optional.ofNullable(finalizeEvent)
+                        .map(event -> ChainUtils.buildChainReceipt(event.log, contribution.getChainTaskId(), getLatestBlockNumber()));
+            }, executor).get();
+        } catch (ExecutionException e) {
+            log.error("contributeAndFinalize asynchronous execution did not complete", e);
+        } catch (InterruptedException e) {
+            log.error("contributeAndFinalize thread has been interrupted", e);
+            Thread.currentThread().interrupt();
+        }
+        return Optional.empty();
+    }
+
+    private IexecHubContract.TaskFinalizeEventResponse sendContributeAndFinalizeTransaction(Contribution contribution, String resultLink, String callbackData) {
+        TransactionReceipt receipt;
+        String chainTaskId = contribution.getChainTaskId();
+
+        RemoteCall<TransactionReceipt> contributeAndFinalizeCall = getWriteableHubContract().contributeAndFinalize(
+                stringToBytes(chainTaskId),
+                stringToBytes(contribution.getResultDigest()),
+                StringUtils.isNotEmpty(resultLink) ? resultLink.getBytes(StandardCharsets.UTF_8) : new byte[0],
+                StringUtils.isNotEmpty(callbackData) ? stringToBytes(callbackData) : new byte[0],
+                contribution.getEnclaveChallenge(),
+                stringToBytes(contribution.getEnclaveSignature()),
+                stringToBytes(contribution.getWorkerPoolSignature()));
+        log.info("Sent contributeAndFinalize [chainTaskId:{}, contribution:{}, resultLink:{}, callbackData:{}]",
+                chainTaskId, contribution, resultLink, callbackData);
+
+        try {
+            receipt = contributeAndFinalizeCall.send();
+            log.debug("Transaction hash {} at block {} [chainTaskId:{}]", receipt.getTransactionHash(), receipt.getBlockNumber(), chainTaskId);
+        } catch (Exception e) {
+            log.error("contributeAndFinalize failed [chainTaskId:{}]", chainTaskId, e);
+            return null;
+        }
+
+        List<IexecHubContract.TaskFinalizeEventResponse> finalizeEvents = getHubContract().getTaskFinalizeEvents(receipt);
+        log.debug("finalizeEvents count {} [chainTaskId:{}]", finalizeEvents.size(), chainTaskId);
+
+        IexecHubContract.TaskFinalizeEventResponse finalizeEvent = null;
+        if (!finalizeEvents.isEmpty()) {
+            finalizeEvent = finalizeEvents.get(0);
+        }
```

Comment:
It would be great to add a comment to explain why we retrieve a list of events but keep only the first one.